File: 2022/typescript/src/day04.ts

function run(input: string): void {
    const res1 = part1(input);
    console.log('Day 04, part 1:', res1);
    const res2 = part2(input);
    console.log('Day 04, part 2:', res2);
}

function part1(input: string): number {
    const lines = input.split('\n');
    const pairs = lines.map(lineToPair);
    const containedPairs = pairs.filter(([f, s]) => contains(f, s));
    return containedPairs.length;
}

function lineToPair(line: string): [[number, number], [number, number]] {
    const [f, s] = line.split(',');
    const [fa, fb] = f.split('-');
    const [sa, sb] = s.split('-');
    return [[Number(fa), Number(fb)], [Number(sa), Number(sb)]];
}

function contains(f: number[], s: number[]): boolean {
    const [fa, fb] = f;
    const [sa, sb] = s;
    if (fa <= sa && fb >= sb) return true;
    if (sa <= fa && sb >= fb) return true;
    return false;
}

function part2(input: string): number {
    const lines = input.split('\n');
    const pairs = lines.map(lineToPair);
    const overlappedPairs = pairs.filter(([f, s]) => overlap(f, s));
    return overlappedPairs.length;
}

function overlap(f: number[], s: number[]): boolean {
    const [fa, fb] = f;
    const [sa, sb] = s;
    // seems odd with three checks, but it works for both the test and real input
    if (fa >= sa && fa <= sb) return true;
    if (fb >= sa && fb <= sb) return true;
    if (sa >= fa && sb <= fb) return true;
    return false;
}

export default { run, part1, part2 };

File: 2022/fsharp/days/Day04.fs
module Day04

let lineToPair (line: string) =
    line.Split(',')
    |> Array.toList
    |> List.map (fun s -> s.Split('-') |> Array.toList |> List.map int)

let contains (points: int list list) =
    match points with
    | [ f; s ] -> f.[0] <= s.[0] && f.[1] >= s.[1] || s.[0] <= f.[0] && s.[1] >= f.[1]
    | _ -> failwithf "Unexpected points: %A" points

let overlap (points: int list list) =
    match points with
    | [ [ fa; fb ]; [ sa; sb ] ] ->
        if fa >= sa && fa <= sb then true
        elif fb >= sa && fb <= sb then true
        elif sa >= fa && sb <= fb then true
        else false
    | _ -> failwithf "Unexpected points: %A" points

let pairs (input: string) (filterFun) =
    let lines = input.Split("\n")
    let pairs = lines |> Array.map lineToPair |> Array.toList
    let filteredPairs = pairs |> List.filter filterFun
    filteredPairs.Length

let part1 (input: string) = pairs input contains

let part2 (input: string) = pairs input overlap

let run (input: string) =
    printfn "Day 04, part1: %A" (part1 input) // 528
    printfn "Day 04, part2: %A" (part2 input) // 881
